From bdb72f20153b8663fc7e5741e985db30e35c6d23 Mon Sep 17 00:00:00 2001
From: jbj <jbj>
Date: Fri, 25 Apr 2014 02:39:00 +0000
Subject: [PATCH 08/46] - sanity.

Signed-off-by: Mark Hatle <mark.hatle@windriver.com>
Signed-off-by: Randy MacLeod <Randy.MacLeod@windriver.com>
---
 rpmio/rpmpgp.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/rpmio/rpmpgp.c b/rpmio/rpmpgp.c
index c795fc5..54bba08 100644
--- a/rpmio/rpmpgp.c
+++ b/rpmio/rpmpgp.c
@@ -65,12 +65,12 @@ pgpImplVecs_t * pgpImplVecs =
     /* implict selection (order DOES matter) */
 #elif defined(WITH_BEECRYPT)
 	&rpmbcImplVecs;
+#elif defined(WITH_GCRYPT)
+	&rpmgcImplVecs;
 #elif defined(WITH_NSS)
 	&rpmnssImplVecs;
 #elif defined(WITH_SSL)
 	&rpmsslImplVecs;
-#elif defined(WITH_GCRYPT)
-	&rpmgcImplVecs;
 #elif defined(WITH_TOMCRYPT)
 	&rpmltcImplVecs;
 #elif defined(WITH_CDSA)
-- 
2.0.1

