From 4eef9ee3aeb60e5e5d3d0a93b33f3efc0cc3ea2f Mon Sep 17 00:00:00 2001
From: jbj <jbj>
Date: Wed, 14 May 2014 22:42:43 +0000
Subject: [PATCH 37/46] - coverity #1214094

Signed-off-by: Mark Hatle <mark.hatle@windriver.com>
Signed-off-by: Randy MacLeod <Randy.MacLeod@windriver.com>
---
 rpmio/rpmpgp.c | 10 ++++++----
 1 file changed, 6 insertions(+), 4 deletions(-)

diff --git a/rpmio/rpmpgp.c b/rpmio/rpmpgp.c
index 5b6f1f5..0ccd580 100644
--- a/rpmio/rpmpgp.c
+++ b/rpmio/rpmpgp.c
@@ -768,7 +768,7 @@ const rpmuint8_t * pgpPrtPubkeyParams(pgpDig dig, const pgpPkt pp,
 		}
 	    }
 	    if (i == 0) {
-		pgpPrtHex(" Curve = [ OID]:", p+1, p[0]);
+		pgpPrtHex("Curve = [ OID]:", p+1, p[0]);
 		pgpPrtNL();
 		p += p[0] + 1;
 	        pgpPrtStr(pgpPublicECDSA[i], pgpMpiStr(p));
@@ -789,12 +789,12 @@ const rpmuint8_t * pgpPrtPubkeyParams(pgpDig dig, const pgpPkt pp,
 		}
 	    }
 	    if (i == 0) {
-		pgpPrtHex(" Curve = [ OID]:", p+1, p[0]);
+		pgpPrtHex("Curve = [ OID]:", p+1, p[0]);
 		p += p[0] + 1;
 		pgpPrtNL();
 	        pgpPrtStr(pgpPublicECDH[i], pgpMpiStr(p));
 		p += pgpMpiLen(p);
-		pgpPrtHex("     KDF params:", p+1, p[0]);
+		pgpPrtHex("    KDF params:", p+1, p[0]);
 		p += p[0] + 1;
 		pgpPrtNL();
 		break;	/* XXX FIXME: early exit */
@@ -1213,8 +1213,10 @@ pgpDigParams sigp = pgpGetSignature(dig);
     int rc = 0;
 
     t = rpmExpand("%{?_build_sign}", NULL);
-    if (!(t && *t))
+    if (!(t && *t)) {
+	t = _free(t);		/* coverity #1214094 */
 	t = xstrdup("DSA");
+    }
     dig->build_sign = t;
 
     if ((te = strrchr(t, '/')) != NULL)
-- 
2.0.1

