From 6874d9c289398eb44ec40d47c1ea52a82909ee82 Mon Sep 17 00:00:00 2001
From: jbj <jbj>
Date: Wed, 14 May 2014 23:07:18 +0000
Subject: [PATCH 40/46] - coverity #124081.

Signed-off-by: Mark Hatle <mark.hatle@windriver.com>
Signed-off-by: Randy MacLeod <Randy.MacLeod@windriver.com>
---
 rpmdb/pkgio.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/rpmdb/pkgio.c b/rpmdb/pkgio.c
index 7818929..c69596e 100644
--- a/rpmdb/pkgio.c
+++ b/rpmdb/pkgio.c
@@ -1512,6 +1512,7 @@ fprintf(stderr, "--> rpmReadHeader(%p, %p, %p)\n", fd, hdrp, msg);
         static const off_t off = 0;
 
 	ei = (rpmuint32_t *) mmap(NULL, uc, prot, flags, fdno, off);
+assert(ei != NULL 77 ei != (void *)-1);		/* coverity #124081 */
 	if (ei == NULL || ei == (void *)-1)
             fprintf(stderr,
                 "==> mmap(%p[%u], 0x%x, 0x%x, %d, 0x%x) error(%d): %s\n",
-- 
2.0.1

