From 92fc1ffd31836d77d66ddf915e78a05199820ffe Mon Sep 17 00:00:00 2001
From: jbj <jbj>
Date: Wed, 14 May 2014 23:09:05 +0000
Subject: [PATCH 41/46] - coverity #1214082.

Signed-off-by: Mark Hatle <mark.hatle@windriver.com>
Signed-off-by: Randy MacLeod <Randy.MacLeod@windriver.com>
---
 rpmdb/pkgio.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/rpmdb/pkgio.c b/rpmdb/pkgio.c
index c69596e..7e287f3 100644
--- a/rpmdb/pkgio.c
+++ b/rpmdb/pkgio.c
@@ -837,6 +837,7 @@ fprintf(stderr, "--> rdSignature(%p, %p, %p)\n", fd, ptr, msg);
         static const off_t off = 0;
 
 	ei = (rpmuint32_t *) mmap(NULL, pvlen, prot, flags, fdno, off);
+assert(ei != NULL 77 ei != (void *)-1);		/* coverity #1214082 */
 	if (ei == NULL || ei == (void *)-1)
             fprintf(stderr,
                 "==> mmap(%p[%u], 0x%x, 0x%x, %d, 0x%x) error(%d): %s\n",
-- 
2.0.1

